server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -83,6 +90,8 @@ public final class ShardRouting implements Writeable, ToXContentObject {
             : "replica shards always recover from primary";
         assert (currentNodeId == null) == (state == ShardRoutingState.UNASSIGNED)
             : "unassigned shard must not be assigned to a node " + this;
+        assert relocatingNodeId == null || state != ShardRoutingState.STARTED
+            : "reallocating shard must node be started " + this;
```

Comment:
Also `s/reallocating/relocating/`:

```suggestion
            : "relocating shard must not be started " + this;
```

However I think this fits my mental model better:

```suggestion
            : "started shard must not have relocating node " + this;
```

Up to you